tests/project_helpers.py

```diff
@@ -0,0 +1,245 @@
+        logger.info(f"Creating venv at {venv_dir}...")
+        venv_script = self.venv_script(venv_dir)
+        subprocess.run(
+            " && ".join(venv_script),
+            check=True,  # fail if any of the commands fail
+            shell=True,  # pass multiple shell commands to the subprocess
+        )
+        # Make sure the venv has been installed
+        assert Path(venv_dir, ".installed").is_file()
+        cache.set(f"fawltydeps/{self.venv_hash()}", str(venv_dir))
+        return venv_dir
+
+
+@dataclass
+class AnalysisExpectations:
+    """Verify our expectations on the analysis resulting from an experiment.
+
+    Encapsulate the expected values of an Analysis object (or its JSON
+    counterpart) after a real/sample project experiment has been run.
+
+    The members of this expectation object are all optional; if omitted, the
+    corresponding Analysis member will _not_ be checked. Furthermore, no "deep"
+    verification of each member is performed here; rather, we only check that
+    the set of (import or dependency) names that can found in each Analysis
+    member match the expected set stored here.
+    """
+
+    imports: Optional[Set[str]] = None
+    declared_deps: Optional[Set[str]] = None
+    undeclared_deps: Optional[Set[str]] = None
+    unused_deps: Optional[Set[str]] = None
+
+    @classmethod
+    def parse_from_toml(cls, data: TomlData) -> "AnalysisExpectations":
+        """Read expectations from the given TOML table."""
+
+        def set_or_none(data: Optional[Iterable[str]]) -> Optional[Set[str]]:
+            return None if data is None else set(data)
+
+        return cls(
+            imports=set_or_none(data.get("imports")),
+            declared_deps=set_or_none(data.get("declared_deps")),
+            undeclared_deps=set_or_none(data.get("undeclared_deps")),
+            unused_deps=set_or_none(data.get("unused_deps")),
+        )
+
+    def _verify_fields(self, member_extractor: Callable[[str], Set[str]]) -> None:
+        for field in fields(self):
+            expected_names = getattr(self, field.name)
+            if expected_names is None:
+                print(f"Skip checking .{field.name}")
+                continue
+
+            print(f"Checking .{field.name}")
+            actual_names = member_extractor(field.name)
+            print(f"  Expect names: {sorted(expected_names)}")
+            print(f"  Actual names: {sorted(actual_names)}")
+            assert expected_names == actual_names
+
+    def verify_analysis(self, analysis: Analysis) -> None:
+        """Assert that the given Analysis object matches our expectations."""
+        self._verify_fields(lambda member: {m.name for m in getattr(analysis, member)})
+
+    def verify_analysis_json(self, analysis: JsonData) -> None:
+        """Assert that the given JSON analysis matches our expectations."""
+        self._verify_fields(lambda member: {m["name"] for m in analysis[member]})
```

Comment:
Ah, the world will be a happier place when we can fully leverage union types and structural pattern matching :)